### ISSUE TITLE
Can't have two identical wait expressions in two script files that call each other.

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#707]
This is a really weird, specific bug that took ages to track down.  I'm sleepy but I want to get the description exactly correct while its still fresh in my head, before I forget it.

**To trigger the bug:**
You have one script that does these things in this order:
  - Lock throttle to anything, even just ```lock throttle to 0.```  (*this probably also happens with lock steering too, but lock throttle is easier to test.**)
  - Then execute a ```wait until <foo>.``` statement, for some expression <foo>, that can be anything.  Just make sure it's a ```wait until``` expression, rather than a raw ```wait <N>.``` to wait N seconds.
  - Then run a second script from within this first script.

You have the second "inner" script that it runs do this:
  - Execute a ```wait until <foo>.``` that uses **precisely the same expression** as the one in the first script.  It's essential that it is the exact same expression, such that it would compile into an identical list of opcodes to evaluate it.
  - that should do it.  It should barf right there.

The error you'll get is "random" (not really random, it is deterministic, but differs depending on stuff not entirely under your control.  It will however be the *same* error each time you run the *same* exact code.)

The error is reported with bogus trace information, and it may cause stack misalignment too.  The actual error is that when the Update() attempts to run through updating the the flight control variables (like the $throttle* lock function) , it will end up executing an OpcodeCall that jumps to some other wrong part of the code instead of into the lock expression.  How many instructions offset the jump is from where it should be will depend on exactly how many instructions it takes to calculate the expression <foo>.  When the <foo> code was inserted up near the top where the trigger conditions go, the jump addresses for the throttle OpcodeCall never got updated to reflect this shift.  The IP location it refers to will be some other part of the program that got shifted down to that location.

This can cause stack misalignments, random re-running of statements that already completed, you name it - thus why the exact error is seemingly "random".

The bug *requires* that the two *wait until* statements be using *exactly* the same expression, or it will not trigger, which is a big clue as to where the problem is.

__Where the problem probably is:__
In the compiler, there appears to be some logic that makes a hashcode number out of the list of opcodes that form the trigger conditional expression, and uses that hashcode to decide if it has already compiled this trigger once before.  Given that the bug only appears when you have the same exact condition in both waits, in two different scripts one that calls the other, that's probably where the problem is.

I originally thought this problem was caused by my functions revamp, because it was my new wrong-nuber-of-args check in OpcodeReturn that was getting thrown.  But after a lot of analysis it turns out that my functions stuff was the "victim" of the problem, not the cause of it.  By jumping a few instructions ahead of where it was supposed to in the code, some of the "pushes" were skipped over, thus making the number of arguments wrong and triggering my message.

This bug was always there, and ziw reports that an error happened even before updating to the functions branch, but it was just a different error message.  (of course, this new exception didn't exist yet, so some other part of the system was offended by the problem of the call to the wrong location).

Here's an example pair of scripts to run that trigger the bug:

file: testbug.ks:
```
// attempt to create the bug with a less complex test.

lock throttle to 0.
lock steering to up.

print "testbug. waiting 2 seconds with a WAIT UNTIL".
set two_seconds_later to time:seconds + 2.
wait until time:seconds > two_seconds_later.
print "calling testbug_first.".
run testbug_first.
print "done".
```

file: testbug_first.ks
```
set aa to 0.5.
set bb to 0.7.
print "RUNNING testbug_first".
print " ".
print "Leaving throttle locked at old value at first.".
print "Waiting 2 seconds with the use of a WAIT UNTIL,".
print "then I will lock throttle to a new value".
set two_seconds_later to time:seconds + 2.
wait until time:seconds > two_seconds_later.
print "wait done, locking throttle.". // probably won't even get this far.
lock throttle to (aa+bb)/2.
print "done with inner".
```

Example of running:
```
run testbug.
testbug. waiting 2 seconds with a WAIT UNTIL
calling testbug_first.
RUNNING testbug_first

Leaving throttle locked at old value at first.
Waiting 2 seconds with the use of a WAIT UNTIL,
then I will lock throttle to a new value
wait done, locking throttle.
done with inner
Cannot cast from source type to destination type.
At testbug on archive, line 7
set two_seconds_later to time:seconds + 2.
                                      ^

```
In the above example the totally bogus message was happening because where the ```lock throttle to 0.``` code still left in place from the first script was trying to jump to, the program code had been shifted down a bit and now in that same memory location was the OpcodeAdd from the expression, and it just jumped right to there and tried running from there, when it should have been jumping a bit further down in the program, to where the ```push 0. return.``` for the ```lock throttle to 0.``` got moved to.

**Comment by Dunbaratu** [comment]
I also just verified that the two duplicate WAIT's must be in *different* scripts to cause the bug.  If they're in the SAME script it's fine, like so:

testbug.ks:
```
// attempt to create the bug with a less complex test.

lock throttle to 0.
lock steering to up.

// wait until, commented out from here:
// print "testbug. waiting 2 seconds with a WAIT UNTIL".
// set two_seconds_later to time:seconds + 2.
// wait until time:seconds > two_seconds_later.
print "calling testbug_first.".
run testbug_first.
print "done".
```

testbug_first.ks:
```
set aa to 0.5.
set bb to 0.7.
// wait until, moved to here instead, where it's in the same file as its duplicate:
print "testbug. waiting 2 seconds with a WAIT UNTIL".
set two_seconds_later to time:seconds + 2.
wait until time:seconds > two_seconds_later.

print "RUNNING testbug_first".
print " ".
print "Leaving throttle locked at old value at first.".
print "Waiting 2 seconds with the use of a WAIT UNTIL,".
print "then I will lock throttle to a new value".
set two_seconds_later to time:seconds + 2.
wait until time:seconds > two_seconds_later.
print "wait done, locking throttle.". // probably won't even get this far.
lock throttle to (aa+bb)/2.
print "done with inner".
```

That variant works fine.  It MUST be that the duplicate waits are in two *different* scripts.

Given what I'm seeing in the compiler code, I wouldn't be surprised if WHEN/THEN and ON have the same exact problem.  It seems to affect all triggers of any type, and a WAIT UNTIL is technically a trigger too which is where it's getting the problem.

**Comment by Dunbaratu** [comment]
grrr. dang it I thought that when you put "Fixes #707" in a title of a PR, it will auto-assocate with issue 707.  That doesn't seem to be working.

This should be closed when PR #710 is merged.
